### PR TITLE
change of the loadFarmers method to recover an error

### PR DESCRIPTION
**Pull request by YvesC88**
I tried to change the loadFarmers method to recover an error and display it later to the user

To check if loadFarmers in the viewmodel is always asynchronous

### CODE REVIEW

File: MonMaraicher/Features/Map/FarmerService.swift [hunk 1]
```diff
@@ -7,21 +7,27 @@
 
 import Foundation
 
+enum LoadFarmerError: Error {
```
Comment (MathisDetourbet) [inline MonMaraicher/Features/Map/FarmerService.swift]: `FarmerServiceError` is more representing your need IMO. Mostly because you won't create a specific Error enum for each func in your service.

Suggestion / tips:

use the namespacing feature to declare your error like that:
```
extension FarmerService {
  enum Error: Swift.Error {....}
 ```
 Because your error only belongs to your service it could be keep inside it. Up to you to do that ofc ;)

File: MonMaraicher/Features/Map/FarmerService.swift [hunk 2]
```diff
@@ -7,21 +7,27 @@
...
+    case invalidJsonFile
+    case errorDecoding
+}
+
 final class FarmerService {
 
-    func loadFarmers() -> [Farmer] {
+    func loadFarmers(handlerError: (Result<[Farmer], LoadFarmerError>) -> Void) {
```
Comment (MathisDetourbet) [inline MonMaraicher/Features/Map/FarmerService.swift]: Suggestion: 
You use Result which totally fine! Nicely done btw! But here inside your func you are using a throwing func by using `try` and `catch`. 
Exercise: make that func throwing and not use Result. You will see the benefits ;)

Comment (MathisDetourbet) [reply MonMaraicher/Features/Map/FarmerService.swift]: The name of the parameter `handleError` is not good. Your are not handling the error only right? It is used also when no error and it gives the data [Farmer]!

Another naming could be `resultHandler` (because your type if Result ;)). Or `completionHandler`...

File: MonMaraicher/Features/Map/MapViewModel.swift [hunk 3]
```diff
@@ -28,8 +28,13 @@ final class MapViewModel: ObservableObject {
     }
 
     private func loadFarmers() {
-        DispatchQueue.main.async {
```
Comment (MathisDetourbet) [inline MonMaraicher/Features/Map/MapViewModel.swift:32]: Why did you delete the DispatchQueue.main? You get the issue you had before...

File: MonMaraicher/Features/Map/FarmerService.swift [hunk 4]
```diff
@@ -7,21 +7,27 @@
...
-               let farmerJson = try String(contentsOfFile: bundlePath).data(using: .utf8) {
-                let farmers = try decoder.decode([Farmer].self, from: farmerJson)
-                return farmers
+            guard let bundlePath = Bundle.main.path(forResource: "Farmers", ofType: "json"),
+                  let farmerJson = try String(contentsOfFile: bundlePath).data(using: .utf8) else {
+                handlerError(.failure(.invalidJsonFile))
+                return
             }
+            let farmers = try decoder.decode([Farmer].self, from: farmerJson)
+            handlerError(.success(farmers))
```
Comment (MathisDetourbet) [inline MonMaraicher/Features/Map/FarmerService.swift]: You can see here, with the bad naming `handleError` it becomes weird to say Im handling error and passing a success 😅

**Review by MathisDetourbet — CHANGES_REQUESTED** [review]
If you need some help tell me ;)
Your proposal wasn't bad at all! I can see you're writing better code now. Keep going 💪

File: MonMaraicher/Features/Map/MapViewModel.swift (hunk 3 above)
Comment (YvesC88) [reply MonMaraicher/Features/Map/MapViewModel.swift:32]: It's not wrong

File: MonMaraicher/Features/Map/FarmerService.swift [hunk 5]
```diff
@@ -7,27 +7,28 @@
...
 
-enum LoadFarmerError: Error {
-    case invalidJsonFile
-    case errorDecoding
-}
-
 final class FarmerService {
 
-    func loadFarmers(handlerError: (Result<[Farmer], LoadFarmerError>) -> Void) {
+    func loadFarmers() throws -> [Farmer] {
```
Comment (MathisDetourbet) [inline MonMaraicher/Features/Map/FarmerService.swift]: Nice!! 👏 
Now you know different error handling options ;)

File: MonMaraicher/Features/Map/MapViewModel.swift [hunk 6]
```diff
@@ -28,12 +28,11 @@ final class MapViewModel: ObservableObject {
...
-            switch result {
-            case .success(let farmers):
-                self.allFarmers = farmers
-            case .failure:
-                print(LoadFarmerError.errorDecoding)
+        DispatchQueue.main.async {
+            do {
+                self.allFarmers = try self.farmerService.loadFarmers()
+            } catch {
+                print("Error decoding: \(FarmerService.Error.decodingError)")
```
Comment (MathisDetourbet) [inline MonMaraicher/Features/Map/MapViewModel.swift]: You are forcing to print the decoding error here... If the error was `invalidJsonFile` how to know it by doing that? Why not printing the error you get with the catch? :) 
The following code should work ;)

```
print("Error when loading farmers: \(error)"
```

**Review by MathisDetourbet — CHANGES_REQUESTED** [review]

File: MonMaraicher/Features/Map/MapViewModel.swift (hunk 6 above)
Comment (YvesC88) [reply MonMaraicher/Features/Map/MapViewModel.swift]: I asked myself the same question when I wrote this code. The JSON file error was not present and it was a mistake IMO
Your suggestion is better because it returns the right error message depending on the case
